test/parallel/test-worker-resource-limits.js

```diff
@@ -0,0 +1,49 @@
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const v8 = require('v8');
+const { Worker } = require('worker_threads');
+
+const resourceLimits = {
+  maxOldSpaceSizeMb: 16,
+  maxSemiSpaceSizeMb: 4,
+  codeRangeSizeMb: 16,
+};
+
+// Do not use isMainThread so that this test itself can be run inside a Worker.
+if (!process.env.HAS_STARTED_WORKER) {
+  process.env.HAS_STARTED_WORKER = 1;
+  const w = new Worker(__filename, { resourceLimits });
+  w.on('exit', common.mustCall((code) => {
+    assert.strictEqual(code, 1);
+  }));
+  w.on('error', common.expectsError({
+    code: 'ERR_WORKER_OUT_OF_MEMORY',
+    message: 'Worker terminated due to reaching memory limit'
+  }));
+  return;
+}
+
+const array = [];
+while (true) {
+  const usedMB = v8.getHeapStatistics().used_heap_size / 1024 / 1024;
+  assert(usedMB < resourceLimits.maxOldSpaceSizeMb);
+
+  let seenSpaces = 0;
+  for (const { space_name, space_size } of v8.getHeapSpaceStatistics()) {
+    if (space_name === 'new_space') {
+      seenSpaces++;
+      assert(space_size / 1024 / 1024 < resourceLimits.maxSemiSpaceSizeMb * 4);
```

Comment:
why * 4 ?

Reply:
@gireeshpunathil Tbh, it’s there so that the test works. In theory, `* 2` should be enough because the new space consists of two semi-spaces, but in practice, V8 seems to consume a bit more memory than what we originally provided.